### PR TITLE
fix: Collator Storage

### PR DESCRIPTION
**Pull request by weichweich**
## Checklist:

- [ ] I have verified that the code works
  - [ ] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

**Review by wischli — APPROVED** [review]
LGTM!

File: pallets/parachain-staking/src/migrations/v5.rs [hunk 1]
```diff
@@ -31,10 +41,30 @@ pub(crate) fn migrate<T: Config>() -> Weight {
 	// Kill selected candidates list
 	old::SelectedCandidates::<T>::kill();
 
-	// count candidates
-	let counter: u32 = CandidatePool::<T>::iter().fold(0, |acc, _| acc.saturating_add(1));
+	// migrate CandidatePool to TopCandidates
```
Comment (Diiaablo95) [inline pallets/parachain-staking/src/migrations/v5.rs:44]: Q: There's no added weight for all these reads/writes that we do now?

File: pallets/parachain-staking/src/migrations/v5.rs [hunk 2]
```diff
@@ -31,10 +41,30 @@ pub(crate) fn migrate<T: Config>() -> Weight {
...
+
+	// copy & count candidates
+	let counter: u32 = old::CollatorState::<T>::iter().fold(0, |acc, (key, candidate)| {
+		CandidatePool::<T>::insert(&key, candidate);
+		acc.saturating_add(1)
+	});
 	CandidateCount::<T>::put(counter);
 
+	// manually kill the ParachainStaking + CollatorState storage.
+	unhashed::kill_prefix(
```
Comment (Diiaablo95) [inline pallets/parachain-staking/src/migrations/v5.rs:63]: C: Just saying, would it be easier to read also afterwards to add the old types to the deprecated module and call `kill` on them? If they are still in the storage, the kill should still take effect. Plus, by having a type defined we can also make sure that in `pre_upgrade` those values exist and in `post_upgrade` that they don't exist anymore.

**Review by Diiaablo95 — COMMENTED** [review]
Overall looks good. I am not entirely sure that deleting the storage by raw hex is the best approach. I would rather retroactively add the types to remove and check in pre and post upgrade they correct deletion.

File: pallets/parachain-staking/src/migrations/v5.rs (hunk 1 above)
Comment (weichweich) [reply pallets/parachain-staking/src/migrations/v5.rs:44]: good point completely forgot the weights...

File: pallets/parachain-staking/src/migrations/v5.rs (hunk 2 above)
Comment (weichweich) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: storagemaps don't have kill. The storage is already in the old `mod`

Comment (Diiaablo95) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: There's a `drain` method, wouldn't that be enough to clear the old storage?

Comment (Diiaablo95) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: I am just saying that we should probably avoid using these hard-coded hex. Even just calling some hex-encoding function on the storage item name would be a better solution.

Comment (wischli) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: I'm not sure `drain` would also kill the prefix. It only drains the entries.

Comment (weichweich) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: don't know who else that works

Comment (weichweich) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: ah drain would be enough. didn't see that!

Comment (weichweich) [reply pallets/parachain-staking/src/migrations/v5.rs:63]: yeah i also don'T like the hardcoded hex

File: pallets/parachain-staking/src/migrations/v5.rs (hunk 1 above)
Comment (weichweich) [reply pallets/parachain-staking/src/migrations/v5.rs:43]: should be fixed

**Review by Diiaablo95 — APPROVED** [review]
🏆

**Pull request merged**